### DISCUSSION TITLE
Testing API: Should I use a single TestController or multiple TestControllers for workspaces?

### DISCUSSION THREAD

**Discussion by JustinGrote** [microsoft/vscode-discussions#687] [Extension Development QnA] [answered]
CC @connor4312 

I'm looking to implement multiple workspace support to my testing extension. The best approach appears to be to run a separate testing process per workspace.

Should I still use a single test controller and do all the coordination at a lower level, or should I do a test controller per workspace? I can't see a strong argument either way other than multiple test controllers seems easier to expose the "root" of each test folder to the top of the testing UI, and I don't see any issues needing to "coordinate" between the different processes since they run more or less independently.

Thoughts?

**Comment by connor4312** [comment, marked as answer]
In the general case having a test controller per workspace folder makes sense, in the event that you have different sets of TestRunProfiles per folder. But this isn't required, and if that doesn't apply to you having a single controller is fine.

**Reply by JustinGrote** [reply]
Thanks, I was more concerned if it was a big nono for a single extension to contribute multiple test controllers for whatever reason. Thanks for clarifying.
